### PR TITLE
Update lakefile with scripts for import generation

### PR DESCRIPTION
**Pull request by 0art0**
This pull request introduces two new scripts to the `lakefile` of `mathlib4`:
- `import_all`: Generate top-level files importing all the files in the libraries of the repository.
- `import_all?`: Check whether the top-level files are up-to-date, which may be useful for CI (this is modelled on [the `lakefile` of `Yatima`](https://github.com/lurk-lab/yatima/blob/main/lakefile.lean#L133-L144), suggested by @arthurpaulino .)

Note that the scripts in this `lakefile` are not inherited by downstream projects with a `mathlib4` dependency.

The scripts can be run by typing `lake run import_all` and `lake run import_all?` respectively.
 
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by semorrison** [comment]
@0art0, just linking to @tydeu's messages so we can track that these issues are addressed here.

https://leanprover.zulipchat.com/#narrow/stream/270676-lean4/topic/Lake.20script.20for.20automatically.20generating.20files.20with.20imports/near/370935304 (`getCurrentPackage` will break when run from another directory)

File: lakefile.lean [hunk 1]
```diff
@@ -53,39 +53,41 @@ section Scripts
...
-
-def importsForLib (lib : Name) : IO String := do
-  let dir ← FilePath.walkDir lib.toString >>= Array.filterM (not <$> ·.isDir)
-  let filePathToImport : FilePath → String := fun fp ↦ fp.toString.takeWhile (· != FilePath.extSeparator) |>.map <|
-    fun c ↦ if c = FilePath.pathSeparator then '.' else c
-  let imports := dir.foldl (init := "") <| fun s f ↦ s ++ s!"import {filePathToImport f}\n"
+def importsForLib (dir : FilePath) (root : Name) : IO String := do
+  let files ← FilePath.walkDir (dir / root.toString) >>= Array.filterM (not <$> ·.isDir)
+  let filePathToImport (fp : FilePath) : String := fp.toString
+    |>.drop dir.toString.length.succ
```
Comment (0art0) [inline lakefile.lean:59]: Is there a more robust way to achieve this? I would like to drop the prefix corresponding to the directory `dir` and retain the remaining portion of the relative file path.

**Comment by 0art0** [comment]
I have updated the script according to @tydeu's suggestions [on Zulip](https://leanprover.zulipchat.com/#narrow/stream/270676-lean4/topic/Lake.20script.20for.20automatically.20generating.20files.20with.20imports).

File: lakefile.lean (hunk 1 above)
Comment (tydeu) [reply lakefile.lean:59]: You cloud use what Lake does for globs (i.e., [`forEachModuleIn`](https://github.com/leanprover/lake/blob/e9745480069793e8276158283bddff968426efa3/Lake/Config/Glob.lean#L26-L35)) or just emulate it.

Comment (0art0) [reply lakefile.lean]: Thanks. I have tried a different approach by creating a function `System.FilePath.makeRelativeTo` that performs the necessary check before modifying the path.

File: lakefile.lean [hunk 2]
```diff
@@ -48,3 +48,53 @@ lean_lib Archive where
 
 lean_lib Counterexamples where
   roots := #[`Counterexamples]
+
+section Scripts
+
+open System
+
+def System.FilePath.makeRelativeTo (dir file : FilePath) : Option FilePath := do
```
Comment (0art0) [inline lakefile.lean:56]: Should this function be moved elsewhere?

Comment (tydeu) [reply lakefile.lean:56]: Personally, I would suggest using `readDir` instead of `walkDir` so you can avoid this post-processing.

**Comment by 0art0** [comment]
Apologies for the delay, there were a few subtle issues in the code (which are now fixed). I have incorporated @tydeu's suggestion of using a variant of `forEachModuleIn` to avoid ad-hoc processing of file names. 

Please let me know if I should run `lake run import_all` and push the results up before this PR is merged. As I mentioned earlier, the order of the imports in `Mathlib.lean` and related top-level files might change - the current order is alphabetical, but the script in this PR orders them according to the depth-first traversal of the file directory.

**Comment by semorrison** [comment]
I think `Mathlib.lean` should remain alphabetical.

**Comment by 0art0** [comment]
I have now made the imports alphabetically-ordered.

**Comment by semorrison** [comment]
I guess if we are going to switch to this, we should replace the `check_imported` section of the `.github/workflows/*.yml` with an invocation of this script?

That would require further modification to allow building `Mathlib/Tactic.lean`.

**Comment by semorrison** [comment]
I'm slightly uncomfortable adding any code at all to the lakefile. This adds to the elaboration time of every `lake build`.

For example, on my system `time (for i in {1..100}; do lake env echo; done)` (i.e. elaborate the lakefile, but do nothing, 100 times) takes 223s on this branch, but on 185s on master.

That means that the mere presence of these scripts in the lakefile is adding about `0.4s` to every single invocation of `lake`, which is unfortunately not viable. Pinging @tydeu in case I'm missing something here.

We could move these scripts into the codebase (e.g. like ImportGraph) for now.

**Comment by 0art0** [comment]
Would moving these scripts to one of the dependencies, say `Std`, be an option? Lake scripts can be invoked in downstream projects (in this case, it will be `lake run std/import_all` instead of `lake run import_all`), as Mac mentioned in a different conversation on Zulip.

**Comment by tydeu** [comment]
@0art0 Not at the moment, Lake elaborates all configuration files (both the root and dependent packages) on each startup. This is something I hope to address soon.

**Comment by 0art0** [comment]
Regarding building `Mathlib/Tactic.lean`, is there a way to make the `Mathlib/Tactic` folder a package in itself? I tried adding variations of
```lean
lean_lib Tactic where
  roots := #[ `«Mathlib/Tactic» ]
```
to the `lakefile`, but haven't had any success yet. If this works, I think `Mathlib/Tactic.lean` will be built automatically by the script each time.

As a follow-up question: even if there is a way to add `Mathlib/Tactic` as a package to the `lakefile`, is this recommended practice?

**Comment by grunweg** [comment]
It seems this PR is superseded by #11849 and #11874 now (which implement the same idea, outside of the lakefile).
In other words, the effect of this is coming to mathlib. Thank you for (perhaps) starting this conversation.

**Pull request closed without merging**